src/reactivity.ts

```diff
@@ -1,277 +1,163 @@
+  }
+  const proxiesForTarget = proxyCache.get(target)!;
+  if (!proxiesForTarget.has(notify)) {
+    const proxy = new Proxy(target, {
+      get(target: any, key: ObjectKey, proxy: Reactive<T>) {
+        if (key === TARGET) {
+          return target;
+        }
+        observeTargetKey(target, key, notify);
+        const value = Reflect.get(target, key, proxy);
+        if (!isObservable(value)) {
+          return value;
+        }
+        return reactive(value, notify);
+      },
+      set(target, key, value, proxy) {
+        const isNewKey = !Object.hasOwnProperty.call(target, key);
+        const originalValue = Reflect.get(target, key, proxy);
+        const ret = Reflect.set(target, key, value, proxy);
+        if (isNewKey) {
+          notifyTarget(target, KEYCHANGES);
+        }
+        // While Array length may trigger the set trap, it's not actually set by this
+        // method but is updated behind the scenes, and the trap is not called with the
+        // new value. We disable the "same-value-optimization" for it because of that.
+        if (originalValue !== value || (Array.isArray(target) && key === "length")) {
+          notifyTarget(target, key);
+        }
+        return ret;
+      },
+      deleteProperty(target, key) {
+        const ret = Reflect.deleteProperty(target, key);
+        notifyTarget(target, KEYCHANGES);
+        notifyTarget(target, key);
+        return ret;
+      },
+      ownKeys(target) {
+        observeTargetKey(target, KEYCHANGES, notify);
+        return Reflect.ownKeys(target);
+      },
+      has(target, key) {
+        // TODO: this observes all key changes instead of only the presence of the argument key
+        observeTargetKey(target, KEYCHANGES, notify);
+        return Reflect.has(target, key);
+      },
+    });
+    proxiesForTarget.set(notify, proxy);
+    if (!callbacksToTargets.has(notify)) {
+      callbacksToTargets.set(notify, new Set());
     }
-    current = current.parent;
-  } while (current);
-  return false;
+    callbacksToTargets.get(notify)!.add(target);
+  }
+  return proxiesForTarget.get(notify) as Reactive<T>;
 }
 
-// set of atoms linked to observers of type ComponentNode
-class Antichain extends Set<Atom> {
-  level?: number;
-  add(atom: Atom) {
-    const node = atom[OBSERVER];
-    if (this.level === node.level || this.level === undefined) {
-      super.add(atom);
-      this.level = node.level;
-      return this;
-    }
-    let willAdd = false;
-    for (const atom2 of this) {
-      const node2 = atom2[OBSERVER];
-      if (!willAdd && isLessOrEqual(node, node2)) {
-        return this;
-      } else if (isLessOrEqual(node2, node)) {
-        super.delete(atom2);
-        willAdd = true;
-      }
+export function batched(callback: callback) {
+  let called = false;
+  return async () => {
+    await Promise.resolve();
+    if (!called) {
+      called = true;
+      callback();
+      await Promise.resolve();
```

Comment:
why wait here?

Reply:
I've added some comments for both `Promise.resolve`, let me know if that clears things up